File: ta4j-core/src/test/java/org/ta4j/core/indicators/numeric/statistics/StandardDeviationIndicatorTest.java
package org.ta4j.core.indicators.numeric.statistics;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.MethodSource;
import org.ta4j.core.MarketEventTestContext;
import org.ta4j.core.api.Indicators;
import org.ta4j.core.num.NumFactory;


class StandardDeviationIndicatorTest {
  private MarketEventTestContext testContext;


  @BeforeEach
  void setUp() {
    this.testContext = new MarketEventTestContext();
    this.testContext.withCandlePrices(1, 2, 3, 4, 3, 4, 5, 4, 3, 0, 9);
  }


  @ParameterizedTest(name = "STDEV [{index}] {0}")
  @MethodSource("org.ta4j.core.NumFactoryTestSource#numFactories")
  void standardDeviationUsingBarCount4UsingClosePrice(final NumFactory numFactory) {
    this.testContext.withNumFactory(numFactory);

    final var sdv = Indicators.closePrice().stddev(4);

    this.testContext.withIndicator(sdv)
        .fastForwardUntilStable()
        .assertCurrent(1.291)
        .assertNext(0.81649)
        .assertNext(0.57735)
        .assertNext(0.81649)
        .assertNext(0.81649)
        .assertNext(0.81649)
        .assertNext(2.1602)
        .assertNext(3.7416);
  }
}

File: ta4j-core/src/test/kotlin/org/ta4j/core/indicators/numeric/statistics/StandardDeviationIndicatorTest.kt
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2017-2024 Ta4j Organization & respective
 * authors (see AUTHORS)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
package org.ta4j.core.indicators.numeric.statistics

import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.MethodSource
import org.ta4j.core.MarketEventTestContext
import org.ta4j.core.api.Indicators
import org.ta4j.core.num.NumFactory

internal class StandardDeviationIndicatorTest {
    private lateinit var testContext: MarketEventTestContext

    @BeforeEach
    fun setUp() {
        testContext = MarketEventTestContext()
        testContext.withCandlePrices(1.0, 2.0, 3.0, 4.0, 3.0, 4.0, 5.0, 4.0, 3.0, 0.0, 9.0)
    }

    @ParameterizedTest(name = "STDEV [{index}] {0}")
    @MethodSource("org.ta4j.core.NumFactoryTestSource#numFactories")
    fun standardDeviationUsingBarCount4UsingClosePrice(numFactory: NumFactory) {
        testContext.withNumFactory(numFactory)

        val sdv = Indicators.closePrice().stddev(4)

        testContext.withIndicator(sdv)
            .fastForwardUntilStable()
            .assertCurrent(1.291)
            .assertNext(0.81649)
            .assertNext(0.57735)
            .assertNext(0.81649)
            .assertNext(0.81649)
            .assertNext(0.81649)
            .assertNext(2.1602)
            .assertNext(3.7416)
    }
}